### ISSUE TITLE
Pass parameter to helper function

### ISSUE THREAD

**Issue by Fauphi** [Urigo/angular-meteor#1087]
How can I pass a parameter to a helper function?

This was possible with Blaze:
```javascript
Template.dummy.helpers({
	getImage: function(imageId) {
		return Images.findOne(imageId);
	}
});
```
```html
{{ getImage '1234' }}
```
How can I do something like that with Angular Meteor?

This is not working: 
```javascript
this.helpers({
	getImage: (imageId) => {
		return Images.findOne(imageId);
	}
});
```

**Comment by MilosStanic** [comment]
You can define `imageId` on the scope, then inside the helper function do:
```javascript
this.imageId = '';

this.helpers({
    getImage: (imageId) => {
        return Images.findOne(this.getReactively('imageId'));
    }
});
```
So, when you change `this.imageId` in the controller, the helper will autorun and apply new value to the parameter of `findOne` inside helper.

**Comment by Fauphi** [comment]
Yes, that would work if the getImage helper is only used on one image. But as soon as there are two or more images using the helper every image is the same.

**Comment by MilosStanic** [comment]
In that case you don't need a helper function, you need an ordinary controller function to get image whenever you need one, as many times as you need. Maybe if you posted how and where you intend to use the `getImage` function, I could understand better and help you.

**Comment by Fauphi** [comment]
It is basically just a ng-repeat with img:

```html
<span class="img-thumbnail" ng-repeat="imgId in workModal.work.images">
    <img class="info-image" ng-src="{{ workModal.getImage(imgId) }}"/>
</span>
```

workModal.work.images is an array of image-IDs. I have to different collections: Work and Images. The Work collection contains items with name, description, image-id-array,... . The Image collection contains items with image-id and image-url.

While I am writing this I found a solution...

The problem was that the Image subscription was not ready as the ng-repeat rendered the img-elements, and because of that a normal controller function couldn't find something in the Image collection. Now I have combined the Image and Work collection into one subscription and it is working.

The whole approach is still a little bit different than in Blaze, have to change my thinking! :)

Thanks Milos for the help!

**Comment by mgrivera** [comment]
Hi @MilosStanic, 

> In that case you don't need a helper function, you need an ordinary controller function ...

Could you please explain when to use a *helper function* and when to use an *ordinary controller function*? 

Aren't helpers, in new version of angular-meteor, supposed to create bindings between the view and your data? 

By the way ... can I use a helper function to bind an ordinary value, I mean, not a minimongo collection (find) or object (findOne)? 

Thanks and bye ...

**Comment by MilosStanic** [comment]
@mgrivera 
Using helper functions is for the so called three-way-binding, i.e. creating reactive resources for your view-model. You use helper functions when you need Meteor.autorun to update the view-model whenever the datasource changes. For example, you create a helper that returns a set of users with their online status, or nicknames, or avatars. You use this helper to populate the view-model and render it in view, for example a list of users which displays their nicknames, avatars and online status. So, now whenever any of those properties changes in the database. For example, one of the loggedin users changes their nickname, or avatar, or goes offline, this change is reflected on the database, and through the reactivity it is reflected on all view-models, and hence all rendered views on all clients connected.

Using a controller function, well, using controller functions is usually reserved to perform actions based on interactions with user. For example: user clicks a button, and then something needs to be changed. There are also other uses, they can be used to call Meteor.methods, or call remote API functions with $http, or whatever. 

Angular, by default, provides two-way binding. It's a binding of view-model with view in both directions. 

This is a very wide topic. I can go on like this forever. Best thing is to follow the tutorials at angular-meteor.com or some other AngularJS tutorials to grasp the difference between Blaze way of thinking and AngularJS way of thinking.

**Comment by Fauphi** [comment]
Hi @MilosStanic ,

thanks for all the explanations and insights! 
But I still have a question: Is it possible to create reactive functions which accept parameters?

I ask this because I still have a problem with my app. If you remember the "Work" collection is used for the ng-repeat:

```html
<li ng-repeat="item in workList.work" class="work-item">
    <img ng-src="{{ workList.getImage(item.mainImage) }}" alt="{{item.description.name}}" ng-init="workList.imageId=item.mainImage" ng-click="workList.showModal(item._id)" />
    </li>
````
```javascript
this.subscribe('allWorkAndMedia'); // subscribes to the complete Work and Images collection

this.helpers({
    work: () => Work.find({})
});
````

While the ng-repeat is rendered it also calls the getImage controller function to get a image-url with a image-id (image-id is saved in the "Work" item).

```javascript
this.getImage = (mediaId) => {
    var obj = Images.findOne(mediaId);
    return obj && obj.url;
}
```

So, the problem is: 
When a "Work" item gets changed in the database everything works fine, because the ng-repeat is rendered again. But if an "Image" item is changed (i.e. the image-url) nothing happens and that is because I can't use a reactive function for the getImage.

Am I missing something? Or is something like that not possible with Angular-Meteor?

**Comment by MilosStanic** [comment]
Hi @Fauphi , 

I'm really not sure I understand what is happening here. 
You have a list of work items and a list of images. So, you render work items, and for each you retrieve an image from Images collection. That is clear. 
So when you change an image that is related to work item, you want it to be re-rendered in work items list. 
I don't know. This is software design issue. Maybe you should consider redesigning relationships and how you keep data. 

For example: I have a list of users, and each has an avatar image which is stored also in Images collection. But I keep a `currentAvatar` field in my users collection. So when I change avatar in Images, I also update the `currentAvatar` collection with new avatar link, and it gets re-rendered in list of users. 

Similarly, you should keep image url for each work item. When you upload/change image for each work item, you update the image url in work items collection. Then it will re-render list with new image. Actually, it will only retrieve a new image when ng-src changes for particular item in work list, because that's how ng-src works. 

I hope I helped.

But, anyway, your scenario is also interesting. If you find a solution to your problem with current design, please post here.

**Comment by MilosStanic** [comment]
A shortcut that might do the trick for you is to define an association between collections of work items and images. 
Please read this: http://guide.meteor.com/collections.html#associations

I am **guessing** that if you have this sort of association, it will reflect reactively to work items list when you change in associated images collection.

**Comment by Fauphi** [comment]
I don't think it is a design issue, just a different design. And it works with Blaze :)

One solution I found is to get the Images already in the helper:

```javascript
    this.helpers({
		work: () => {
			var workArray = Work.find({}).fetch();

			if(workArray) {
				for(var i=workArray.length;i--;) {
					workArray[i].mainImageUrl = Media.findOne({_id: workArray[i].mainImage}).url;
				}
				return workArray;
			}

		}
	});
```
This works fine and is basically the same as my previous approach, only combined in one function.

Just out of interest, is there a specific reason why a helper function doesn't accept any parameters?

Thanks for the tip with the collection helpers, this looks very promising!

**Comment by mgrivera** [comment]
Hi @Fauphi, I'm just digesting the syntax of the new version and comparing to the one we have had until now. 

When you say helper functions don't accept parameters, which do you refer to? Look at this code from the tutorials: 
```javascript
angular.module('socially').directive('partiesList', function () {
  return {
    restrict: 'E',
    templateUrl: 'parties-list.html',
    controllerAs: 'partiesList',
    controller: function ($scope, $reactive) {
      $reactive(this).attach($scope);
 
      this.newParty = {};
 
      this.helpers({
        parties: () => {
          return Parties.find({});
        }
      });
 
      this.addParty = () => {
        Parties.insert(this.newParty);
        this.newParty = {};
      };
 
      this.removeParty = (party) => {
        Parties.remove({_id: party._id});
      }
    }
  }
});
```
the function *removeParty* is accepting *party* as a parameter ...

**Comment by mgrivera** [comment]
Wow, now I know what you are talking about. The function *removeParty* is not part of helpers ...

**Comment by Fauphi** [comment]
Hi @mgrivera ,

with helper functions I mean those inside this.helpers({}). In your example that is only "parties".

The removeParty function is a normal controller function which is not reactive.

**Comment by gaykan** [comment]
Hi @Fauphi, @mgrivera,
using this tools in real life applications, dealing with complex structures, other than users and simple additional data  needs some more guidelines. We are after 3 way bindings. But what will be the limit?

Even for this parties application, if there are large number of users, will holding a local copy of users data be feasible and possible just to see the avatar changing live. If you have more data tables !

So dear @Fauphi, if you do not have a way of just adding the image array for only those users in the viewer screen , not for the all would be a better design.

We need to make a sample application, and not just only rely on this starter demo.

Would like to here more from your experience...





.

**Comment by MilosStanic** [comment]
@gaykan 
Yes, you're absolutely right. 
Real world application of angular-meteor goes well beyond samples that we have been able to see on the main site. 
Display of users list is a particular problem that took me a couple of weeks solving until someone helped me here in github issues. 
Also, getting url's from freshly uploaded images presented a separate problem until I discovered there's a 'bug' in CollectionFS and worked around the problem.
Reading up on some plain Meteor articles on best practices regarding publishing/subscribing may help.
As a conclusion -- there's always a trade-off between features brought to us by framework that make our lives as programmers easy, and complications that come together with features. In sum, angular-meteor brings more benefits for me.
